### PR TITLE
1349 automate elastic search using terraform

### LINKED ISSUE #1349: Terraform: Automate Deployment of Elasticsearch
**Description:**
For each Cloud.gov space we need to update the Terraform configuration files to deploy the [AWS Elasticsearch](https://cloud.gov/docs/services/aws-elasticsearch/) service. The plan used should be `es-dev` for `tanf-dev` and `tanf-staging`, while `tanf-prod` should use the `es-medium-ha` plan.

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] Terraform automatically provisions Elasticsearch for each Cloud.gov space
- [ ] Appropriate plans are used per space
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ] Update Terraform configuration files to deploy Elasticsearch to Cloud.gov
- [ ] Bind the deployed service to any app in Cloud.gov and confirm it can be reached via cURL from the Cloud.gov app (ssh)
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**

**Supporting Documentation:**

**Open Questions:**

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes

Pull request closes #1349 _

## How to Test

1. change directory to terraform/dev/
1. run ``` terraform plan``` : this will show the conf that will deploy
1. (optional) you can also try to deploy. NOTE: this will change terraform instance on cloud.gov/dev
## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] [**_insert ACs here_**]
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: terraform/staging/main.tf [hunk 1]
```diff
@@ -76,3 +76,13 @@ resource "cloudfoundry_service_instance" "datafiles" {
...
 }
+
+data "cloudfoundry_service" "elasticsearch" {
+  name = "aws-elasticsearch"
+}
+
+resource "cloudfoundry_service_instance" "elasticsearch" {
+  name             = "es-staging"
+  space            = data.cloudfoundry_space.space.id
+  service_plan     = data.cloudfoundry_service.elasticsearch.service_plans["es-dev"]
```
Comment (raftmsohani) [inline terraform/staging/main.tf:87]: need to change the plan on this one

File: terraform/production/main.tf [hunk 2]
```diff
@@ -76,3 +76,13 @@ resource "cloudfoundry_service_instance" "datafiles" {
...
 }
+
+data "cloudfoundry_service" "elasticsearch" {
+  name = "aws-elasticsearch"
+}
+
+resource "cloudfoundry_service_instance" "elasticsearch" {
+  name             = "es-prod"
+  space            = data.cloudfoundry_space.space.id
+  service_plan     = data.cloudfoundry_service.elasticsearch.service_plans["es-dev"]
```
Comment (elipe17) [inline terraform/production/main.tf]: Should the plan be `.service_plans["es-medium"]`? I think we use medium in prod?

**Review by elipe17 — APPROVED** [review]
Just had a question about the plan in prod, otherwise looks good!

**Comment by raftmsohani** [comment]
> Just had a question about the plan in prod, otherwise looks good!

right,I've already changed that!

**Review by jtimpe — APPROVED** [review]

File: terraform/staging/main.tf (hunk 1 above)
Comment (ADPennington) [reply terraform/staging/main.tf:87]: @raftmsohani does that need to be done here or in a future ticket?

**Review by ADPennington — APPROVED** [review]
👍🏾 ready to merge

**Pull request merged**